docs/pages/technical-specs/expo-updates-0.md

```diff
@@ -81,6 +84,7 @@ content-type: *
 * `expo-server-defined-headers` is an [Expo SFV](expo-sfv.md) dictionary. It defines headers that a client library MUST store until overwritten by a newer dictionary, and they MUST be included in every subsequent [manifest request](#manifest-request).
 * `cache-control` - A value of `cache-control: private, max-age=0` is recommended to ensure the newest manifest is returned. Setting longer cache ages could result in stale updates.
 * `content-type` MUST be determined by _proactive negotiation_ as defined in [RFC 7231](https://tools.ietf.org/html/rfc7231#section-3.4.1). Since the client library is [required](#manifest-request) to send an `accept` header with each manifest request, this will always be either `application/expo+json`, `application/json`; otherwise the request would return a `406` error.
+* `expo-signed-hash` MAY contain the sha256 digest of the manifest encrypted using the [codesigning](#codesigning) private key.
```

Comment:
I still need to look more into the feasibility of arbitrary signature algorithms. It seems like it should be possible but may open up a class of bugs where certain platforms and platform versions don't support particular algorithms, which would then have negative effects on end users. For example, on android the table is pretty complex: https://developer.android.com/reference/java/security/Signature

Reply:
Looked into this more concretely (which is out of scope of this PR but is necessary to determine feasibility). 

Notes:
- The signature (and signature algorithm) of the certificate doesn't necessarily mean that the private key can't be used with different digest algorithms. For example, a cert signed with `sha256WithRSAEncryption` will have a public key that is just an RSA public key, which could be used to verify a `sha512WithRSAEncryption` signature as long as it is known ahead of time that the data was signed with that algorithm. It's unclear to me whether it is common practice for the signature algorithm of the certificate to dictate the signature algorithm of the responses. It seems like it may be more common for the information indicating the algorithm that was used to do the signing to be included in metadata stored with the key and associated with the key. (example draft https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-12#section-4.1)
- In Mac Keychain Access, the only algorithms I can use are RSA and elliptic-curve (ecdsa) for the key pair, and it seems that it only hashes the certificate during the signing process with sha256.
- Using EC in node seems to be prohibitively difficult as neither crypto nor node-forge support it very well. Just a data point to help us consider whether it makes sense to only support RSA since writing servers in node to use EC is tough (not that our internal implementations would need to use EC).
- On android it seems like it's easy to do in code (just feed the certificate's `sigAlgName` directly into `Signature.getInstance`). I'm still concerned about varying API level support for different algorithms causing issues for end-users, so I'm tempted to limit it to really old ones (API 11+ in the chart at https://developer.android.com/reference/java/security/Signature).
- On iOS, allowing arbitrary signature algorithms seems like it will be a lot more difficult to do (many hundreds of LOCs vs 1 LOC in android). Essentially we'd need to [parse the certificate manually, extract out it's signature algorithm field](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/certificates/getting_certificate_values?language=objc), [manually map that to the signature algorithm constants used in the signature verification method](https://stackoverflow.com/questions/61150265/how-do-i-parse-an-x509-certificate-and-extract-its-keys-signature-algorithm), than then use [`SecKeyVerifySignature`](https://developer.apple.com/documentation/security/1643715-seckeyverifysignature?language=objc) with that algorithm constant.

My suggestion here would be to support just `sha256WithRSAEncryption` for now in our implementations so that we don't have to dig into the certificate manually on iOS. We'll want to be sure that we don't limit ourselves to never supporting others in the future though. And we'll also want to be sure that if someone follows this spect they don't have to use any one in particular.

Future support could look like:
- Two fields in client configuration: `cert` and `asset-signing-metadata`, where `cert` is the certificate (which contains the public key and the type of algorithm used [rsa, ec]) and `asset-signing-metadata` contains information about the digest that should be used.
- On Android, we use the metadata in combination with the public key algorithm to determine what signature implementation should be used. Can keep this in a simple hardcoded lookup table.
- On iOS, this is much more difficult since we'll need to dig into the cert to get the public key algorithm. That being said, only supporting a subset might cut down complexity by a bit. We could also duplicate the algorithm into the metadata, but this risks inconsistencies.

---

So, how does this apply to this PR?

In this PR, we need to decide the format of the `expo-signature` header, and also need to decide whether to prescribe a limit in the spec to what algorithms clients and servers need to use. My gut says that the spec shouldn't prescribe the algorithms since an implementor could hardcode their implementations to use any algorithm they like and the spec would still work.

Using the format of https://tools.ietf.org/id/draft-cavage-http-signatures-11.html#rfc.section.2.1 for the signature header seems a bit heavyweight since we don't need any of the parameters except the signature itself (we only have one key on the device so the parameters can't help us). Including `algorithm` in the params was deprecated since it caused a [security issue](https://github.com/spacemonkeygo/httpsig/issues/7#issuecomment-670947824), so algorithm needs to be hardcoded on the client (either by using metadata approach above or by just using the same as the certificate signing algorithm).

Reply:
Looked into this more concretely (which is out of scope of this PR but is necessary to determine feasibility). 

Notes:
- The signature (and signature algorithm) of the certificate doesn't necessarily mean that the private key can't be used with different digest algorithms. For example, a cert signed with `sha256WithRSAEncryption` will have a public key that is just an RSA public key, which could be used to verify a `sha512WithRSAEncryption` signature as long as it is known ahead of time that the data was signed with that algorithm. It's unclear to me whether it is common practice for the signature algorithm of the certificate to dictate the signature algorithm of the responses. It seems like it may be more common for the information indicating the algorithm that was used to do the signing to be included in metadata stored with the key and associated with the key. (example draft https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-12#section-4.1)
- In Mac Keychain Access, the only algorithms I can use are RSA and elliptic-curve (ecdsa) for the key pair, and it seems that it only hashes the certificate during the signing process with sha256.
- Using EC in node seems to be prohibitively difficult as neither crypto nor node-forge support it very well. Just a data point to help us consider whether it makes sense to only support RSA since writing servers in node to use EC is tough (not that our internal implementations would need to use EC).
- On android it seems like it's easy to do in code (just feed the certificate's `sigAlgName` directly into `Signature.getInstance`). I'm still concerned about varying API level support for different algorithms causing issues for end-users, so I'm tempted to limit it to really old ones (API 11+ in the chart at https://developer.android.com/reference/java/security/Signature).
- On iOS, allowing arbitrary signature algorithms seems like it will be a lot more difficult to do (many hundreds of LOCs vs 1 LOC in android). Essentially we'd need to [parse the certificate manually, extract out it's signature algorithm field](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/certificates/getting_certificate_values?language=objc), [manually map that to the signature algorithm constants used in the signature verification method](https://stackoverflow.com/questions/61150265/how-do-i-parse-an-x509-certificate-and-extract-its-keys-signature-algorithm), than then use [`SecKeyVerifySignature`](https://developer.apple.com/documentation/security/1643715-seckeyverifysignature?language=objc) with that algorithm constant.

My suggestion here would be to support just `sha256WithRSAEncryption` for now in our implementations so that we don't have to dig into the certificate manually on iOS. We'll want to be sure that we don't limit ourselves to never supporting others in the future though. And we'll also want to be sure that if someone follows this spect they don't have to use any one in particular.

Future support could look like:
- Two fields in client configuration: `cert` and `asset-signing-metadata`, where `cert` is the certificate (which contains the public key and the type of algorithm used [rsa, ec]) and `asset-signing-metadata` contains information about the digest that should be used.
- On Android, we use the metadata in combination with the public key algorithm to determine what signature implementation should be used. Can keep this in a simple hardcoded lookup table.
- On iOS, this is much more difficult since we'll need to dig into the cert to get the public key algorithm. That being said, only supporting a subset might cut down complexity by a bit. We could also duplicate the algorithm into the metadata, but this risks inconsistencies.

---

So, how does this apply to this PR?

In this PR, we need to decide the format of the `expo-signature` header, and also need to decide whether to prescribe a limit in the spec to what algorithms clients and servers need to use. My gut says that the spec shouldn't prescribe the algorithms since an implementor could hardcode their implementations to use any algorithm they like and the spec would still work.

Using the format of https://tools.ietf.org/id/draft-cavage-http-signatures-11.html#rfc.section.2.1 for the signature header seems a bit heavyweight since we don't need any of the parameters except the signature itself (we only have one key on the device so the parameters can't help us). Including `algorithm` in the params was deprecated since it caused a [security issue](https://github.com/spacemonkeygo/httpsig/issues/7#issuecomment-670947824), so algorithm needs to be hardcoded on the client (either by using metadata approach above or by just using the same as the certificate signing algorithm).

Reply:
> I would also like to leave the door open for others. It's also OK if the algorithm is supported only on one platform or certain devices: for example, devices could specify what signature algorithm they prefer and the server would try to cater to those preferences, returning no signature if it can't.

The most recent HTTP signatures proposal has a specification for an "Accept-Signature" header which may contain an `alg` parameter: https://www.ietf.org/archive/id/draft-ietf-httpbis-message-signatures-06.html#name-the-accept-signature-field

I'm not opposed to having something like this but I think we need to think through the concrete use case:
1. Developer creates key pair 1 (KP1) `[KP1.private, KP1.public]` for code signing.
2. Embed `KP1.public` in app version 1. All version 1 binaries require all responses to be signed with `KP1.private`.
3. Developer rotates keys to start using a better algorithm. Generates KP2 `[KP2.private, KP2.public]`.
4. Replace embedded public key in app version 2 with `KP2.public`. All version 2 binaries require all responses to be signed with `KP2.private`.
5. So now there are requests coming in to the manifest server from both binary version 1 and version 2, and the server needs to know what key to use to sign the response.

Some considerations:
- There are many reasons a developer could rotate keys: new algorithm, lost private key, same algorithm but new key owner (still have old key). Thus there needs to be a way to uniquely identify keys and for that identifier to be sent in the accept-signature header as well as in the response. 
- There is a case for the server to send multiple signatures. In the case above, it may be simpler to just always sign with `KP1.private` and `KP2.private` and put them both in the signature field. Thus the signature field should be able to specify multiple signatures. See https://www.ietf.org/archive/id/draft-ietf-httpbis-message-signatures-06.html#name-multiple-signatures
- There may be a case for the client to have multiple embedded public keys, but I'm having trouble thinking of one.
- There's probably an argument that key rotation adds unnecessary complexity and it makes more sense for the old binaries to just not receive updates.

> I'd like to understand the attack better. As I understand it, the security issue with JWTs was that an attacker could make a token that said it used a no-op signing algorithm and compliant servers would say, "Yep, this JWT is correctly unsigned!" which makes everything useless. Was this the same issue with the HTTP signatures spec?

From my understanding, the attack was something like: Attacker is somewhere in the middle, sets the algorithm to something very weak and easily crackable, creates a signature for the (public key, maliciously mutated data) pair using the weak algorithm, and then sends the response. The client then uses the weak algorithm and signature and verifies the data. That being said I feel like I'm still missing something as it's not easy to reason about the feasibility of such an attack.